File: 백준/Gold/16236. 아기 상어/아기 상어.py
import sys
input = sys.stdin.readline
from heapq import heappush, heappop

n = int(input())
space = [list(map(int, input().split())) for _ in range(n)]
size = 2
eat = 0
start = []
dx = [0, 1, 0, -1]
dy = [-1, 0, 1, 0]

def bfs(x, y):
    global size, eat
    visited = [[0]*n for _ in range(n)]
    heap = []
    space[x][y] = 0
    heappush(heap, (0, x, y))
    while heap:
        depth, x, y = heappop(heap)
        if space[x][y] != 0 and space[x][y] < size:
            eat += 1
            space[x][y] = 0
            start.append((x,y))
            if eat == size:
                eat = 0
                size += 1
            return visited[x][y]
        for i in range(4):
            nx, ny = x + dx[i], y + dy[i]
            if 0 > nx or 0 > ny or n <= nx or n <= ny:
                continue
            if visited[nx][ny] or space[nx][ny] > size:
                continue
            heappush(heap, (depth+1, nx, ny))
            visited[nx][ny] = depth + 1
    return 0

for i in range(n):
    for j in range(n):
        if space[i][j] == 9:
            start.append((i,j))

time = 0
for i,j in start:
    time += bfs(i,j)
print(time)

File: 백준/Gold/16236. 아기 상어/아기 상어.js
const input = require('fs').readFileSync(process.platform === 'linux' ? '/dev/stdin' : __dirname + '/input.txt').toString().split('\n')

const n = Number(input[0])
const board = []
for (let i = 1; i < n + 1; i++) {
  board.push(input[i].split(' ').map(Number))
}

const candidates = []
const shark = { x: -1, y: -1, size: 2, eat: 0 }
for (let i = 0; i < n; i++) {
  for (let j = 0; j < n; j++) {
    if (!board[i][j]) continue
    if (board[i][j] === 9) {
      shark.x = i
      shark.y = j
      board[i][j] = 0
      continue
    }

    candidates.push({ x: i, y: j, size: board[i][j]})
  }
}
candidates.sort((a, b) => {
  if (a.size === b.size) {
    if (a.x === b.x) return a.y - b.y
    return a.x - b.x
  }

  return a.size - b.size
})

const directions = [[0, 1], [0, -1], [1, 0], [-1, 0]]
const time = move_shark()
console.log(time)

function move_shark() {
  let time = 0

  while(true) {
    const fish = find_fish()
    if (!fish) return time
    
    const { distance, x, y } = fish
    time += distance
    shark.eat++
    board[x][y] = 0
    
    if (shark.eat === shark.size) {
      shark.eat = 0
      shark.size++
    }
    
    shark.x = x
    shark.y = y
  }
}

function find_fish() {
  const fish = []
  const visited = Array.from({ length: n }, () => Array.from({ length: n }, () => -1))
  let idx = 0
  const queue = []
  
  visited[shark.x][shark.y] = 0
  queue.push([shark.x, shark.y])

  while(idx < queue.length) {
    const [x, y] = queue[idx++]

    if (board[x][y] && board[x][y] < shark.size) {
      fish.push({ distance: visited[x][y], x, y })
    }

    for (const [dx, dy] of directions) {
      const [nx, ny] = [dx + x, dy + y]

      if (nx < 0 || ny < 0 || nx >= n || ny >= n) continue
      if (visited[nx][ny] !== -1) continue
      if (shark.size < board[nx][ny]) continue

      visited[nx][ny] = visited[x][y] + 1
      queue.push([nx, ny])
    }
  }

  fish.sort((a, b) => {
    if (a.distance === b.distance) {
      if (a.x === b.x) return a.y - b.y
      return a.x - b.x
    }

    return a.distance - b.distance
  })

  return fish[0]
}
